### PR TITLE
Feature/talkol/benchmark consensus continued2

### PR DESCRIPTION
**Pull request by talkol**

### CODE REVIEW

File: services/consensusalgo/benchmarkconsensus/leader.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+			s.reporting.Error(err)
+		}
+		select {
+		case <-ctx.Done():
+			s.reporting.Info("Consensus round run loop terminating with context")
+			return
+		case s.lastSuccessfullyVotedBlock = <-s.successfullyVotedBlocks:
+			s.reporting.Infof("Consensus round waking up after successfully voted block %d", s.lastSuccessfullyVotedBlock)
+			continue
+		case <-time.After(time.Duration(s.config.BenchmarkConsensusRoundRetryIntervalMillisec()) * time.Millisecond):
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/leader.go]: great naming :-)

File: services/consensusalgo/benchmarkconsensus/leader.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		BlockProof:          nil,
+	}
+	blockPair, err := s.leaderSignBlockProposal(transactionsBlock, resultsBlock)
+	if err != nil {
+		panic(err)
+	}
+	return blockPair
+}
+
+func (s *service) leaderGenerateNewProposedBlockUnsafe() (*protocol.BlockPairContainer, error) {
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/leader.go]: why 'unsafe'?

File: services/consensusalgo/benchmarkconsensus/leader.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		s.reporting.Error(err) // TODO: wrap with added context
+		return
+	}
+
+	// add the vote
+	s.lastCommittedBlockVoters[sender.SenderPublicKey().KeyForMap()] = true
+
+	// count if we have enough votes to move forward
+	existingVotes := len(s.lastCommittedBlockVoters) + 1
+	neededVotes := int(math.Ceil(float64(s.config.NetworkSize(0)) * 2 / 3))
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/leader.go]: extract this calculation to a function and name according to meaning of the calc ("requiredQuorumSize"?)

File: services/consensusalgo/benchmarkconsensus/test/leader_test.go [hunk 4]
```diff
@@ -6,11 +6,121 @@ import (
 	"testing"
 )
 
-func TestLeaderProposesBlocks(t *testing.T) {
+func newLeaderHarnessAndInit(t *testing.T, ctx context.Context) *harness {
+	h := newHarness(true)
+	h.expectNewBlockProposalNotRequested()
+	h.expectCommitSent(0, h.config.NodePublicKey())
+	h.createService(ctx)
+	h.verifyCommitSent(t)
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/test/leader_test.go]: why do we verify that commit was sent on init?

File: services/consensusalgo/benchmarkconsensus/test/leader_test.go [hunk 5]
```diff
@@ -6,11 +6,121 @@ import (
...
+	})
+}
+
+func TestLeaderCommitsNextBlockAfterEnoughConfirmations(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		h := newLeaderHarnessAndInit(t, ctx)
+
+		h.expectNewBlockProposalRequestedAndSaved(1)
+		h.expectCommitSent(1, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(3, 0, true, true)
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/test/leader_test.go]: it's very hard to understand the meaning of these 4 arguments from reading the code. Maybe pass an argument struct so we can name the parameters?

Comment (electricmonk) [reply services/consensusalgo/benchmarkconsensus/test/leader_test.go]: or even split to multiple functions, or a builder

File: services/consensusalgo/benchmarkconsensus/test/leader_test.go [hunk 6]
```diff
@@ -6,11 +6,121 @@ import (
...
+
+		h.expectNewBlockProposalRequestedAndSaved(2)
+		h.expectCommitSent(2, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(3, 1, true, true)
+		h.verifyNewBlockProposalRequestedAndSaved(t)
+		h.verifyCommitSent(t)
+	})
+}
+
+func TestLeaderRetriesCommitOnError(t *testing.T) {
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/test/leader_test.go]: how does this this test assert that the leader retries? I don't see anything related to retries in the code

File: services/consensusalgo/benchmarkconsensus/test/leader_test.go [hunk 7]
```diff
@@ -6,11 +6,121 @@ import (
...
+
+		h.expectNewBlockProposalRequestedToFail()
+		h.expectCommitNotSent()
+		h.receivedCommittedViaGossipFromSeveral(3, 0, true, true)
+		h.verifyNewBlockProposalRequestedAndNotSaved(t)
+		h.verifyCommitNotSent(t)
+	})
+}
+
+func TestLeaderRetriesCommitAfterNotEnoughConfirmations(t *testing.T) {
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/test/leader_test.go]: I don't understand how this test name relates to the body

File: services/consensusalgo/benchmarkconsensus/test/leader_test.go [hunk 8]
```diff
@@ -6,11 +6,121 @@ import (
...
+
+func TestLeaderIgnoresBadSignatures(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		h := newLeaderHarnessAndInit(t, ctx)
+
+		h.expectNewBlockProposalNotRequested()
+		h.expectCommitSent(0, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(3, 0, false, true)
+		h.verifyNewBlockProposalNotRequested(t)
+		h.verifyCommitSent(t)
```
Comment (electricmonk) [inline services/consensusalgo/benchmarkconsensus/test/leader_test.go]: why do we verify that commit was sent? shouldn't we ignore the bad signatures?

Comment (electricmonk) [reply services/consensusalgo/benchmarkconsensus/test/leader_test.go]: or is this an assertion on the gossip messages? if so, why are you even asserting on that?

**Comment by talkol** [comment]
Closed PR since this PR is included in my next PR:

https://github.com/orbs-network/orbs-network-go/tree/feature/benchmark-consensus-handlers

File: services/consensusalgo/benchmarkconsensus/leader.go (hunk 2 above)
Comment (talkol) [reply services/consensusalgo/benchmarkconsensus/leader.go]: Unsafe means that it doesn't capture mutex and assumes the mutex is captured. Meaning if you don't have the mutex, don't call an unsafe function

File: services/consensusalgo/benchmarkconsensus/leader.go (hunk 1 above)
Comment (talkol) [reply services/consensusalgo/benchmarkconsensus/leader.go]: Really?

File: services/consensusalgo/benchmarkconsensus/leader.go (hunk 2 above)
Comment (talkol) [reply services/consensusalgo/benchmarkconsensus/leader.go]: Rename to "UnderMutex"

**Pull request closed without merging**